src/details/hook_rooting.h

```diff
@@ -0,0 +1,254 @@
+//
+// The call to hook methods through kaleidoscope::Hooks and kaleidoscope::HookLoop
+// is templated to allow for compile time static polymorphisms
+// (hook methods of plugins not beeing virtual).
+//
+// ***For non template experts:
+//       This approach is somewhat similar to duck typing in scripting languages
+//       like, e.g. Python. Only the signature of a method counts and not
+//       the fact that the class
+//       is actually derived from a specific base class. Strictly speaking,
+//       plugins do not need to be derived from kaleidoscope::Plugin, but
+//       only need to implement a set of hooks whose call signature matches.
+//
+// If a plugin implements a hook method (thereby hiding the equivalent method
+// of the plugin base class) the derived plugins's method is called instead
+// of the version of the hook method provided by base class kaleidoscope::Plugin.
+// This is possible as the hook method is not invoked via a base class ptr
+// or reference but via a reference to the actual derived plugin class.
+// Thus the compiler can see the actual type of the plugin at the point of
+// the hook method's invocation.
+//
+// HookLoop::apply() implements a compile time for-each loop
+// over all plugins. Under the assumption that only some (few) plugins
+// implement a greater number of hook methods and that there is only a limited number
+// of plugins used in a sketch, this approach is quite efficient both in terms
+// of code size and run-time. This is due to the compiler
+// optimizing away any calls to hooks that have not been
+// implemented by plugins because the base class versions of these hooks are
+// mostly noops. The latter are easy to detect and eliminate by the compiler.
+//
+// Calling the plugins' hook methods is carried out via a call to
+// one of the static functions of class Hooks. Those functions cast the
+// related hook method on every registered plugin by using the HookLoop helper
+// class.
+//
+// Some hooks such as e.g. the eventHandlerHook require a decision about
+// whether to continue calling other hook methods after them, or,
+// if required, to abort the loop over the registered plugins.
+// This decision is implemented through predicated ContinuationPredicate__,
+// a helper class
+// whose eval-method generates a boolean return value that is computed
+// based on the hooks' return values. If the predicate's result is true, the loop
+// continues and is aborted otherwise. The eventHandlerHook method's
+// return value, e.g., signals by itself whether it wants to allow other
+// eventHandlerHooks to be called after it.
+// By means of this approach for most hooks the decision is evaluated at
+// compile-time, which e.g. causes the decision to be optimized out
+// for all those hook methods that have a void return value.
+
+// ReturnTypeTraits is an auxiliary traits class that determines the
+// return type of a hook member function.
+//
+template<typename T>
+struct ReturnTypeTraits {};
```

Comment:
Traits classes are needed to determine specific properties of C++ types at compile time. They are an essential part of TMP. Here we determine the return type of a specific hook method.